File: c/examples/identity.c
/*
 * identity.c — print the Identity object of a CIP device.
 *
 *   1. bp_client_get_id_local()    — LOCAL cm1756 (no path)
 *   2. bp_client_get_device_id()   — REMOTE by text path, OEM-parsed
 *      (and dumps the active node table for context)
 *
 * For per-slot Identity via raw UCMM, see msgprobe with
 * --req "01 02 20 01 24 01" (Get_Attributes_All on Identity).
 *
 * Usage:
 *   identity                    # local cm1756
 *   identity --path "P:1,S:1"   # remote via OEM-parsed text path
 *   identity --path "P:1,S:2"
 *
 * SPDX-License-Identifier: MIT
 */
#include <getopt.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "bpclient.h"

static const char *vendor_name(uint16_t id) {
    switch (id) {
        case 0x0001: return "Allen-Bradley (Rockwell)";
        case 0x0030: return "Online Development Inc.";
        default:     return "(unknown)";
    }
}

static void print_id(const bp_id_object_t *id) {
    /* product_name in OCXcip_GetIdObject responses is sometimes a
     * "SHORT_STRING"-padded buffer (one-byte length prefix, then the
     * bytes); other times the length is implicit and the buffer is
     * just zero-padded.  Print up to 31 bytes ignoring leading 0x00
     * NUL prefix bytes and trailing NULs. */
    char nm[33] = {0};
    size_t skip = 0;
    while (skip < 31 && id->product_name[skip] == 0) skip++;
    size_t take = 0;
    while (skip + take < 32 && id->product_name[skip + take] != 0
                            && take < 32) {
        nm[take] = (char)id->product_name[skip + take];
        take++;
    }
    nm[take] = 0;
    printf("  Vendor ID       : 0x%04x  (%s)\n", id->vendor_id, vendor_name(id->vendor_id));
    printf("  Device Type     : 0x%04x\n", id->device_type);
    printf("  Product Code    : 0x%04x  (%u)\n", id->product_code, id->product_code);
    printf("  Revision        : %u.%u\n", id->major_rev, id->minor_rev);
    printf("  Status          : 0x%04x\n", id->status);
    printf("  Serial Number   : 0x%08x  (%u)\n", id->serial_number, id->serial_number);
    printf("  Product Name    : '%s'\n", nm);
}

int main(int argc, char *argv[]) {
    const char *text_path = NULL;
    static struct option opts[] = {
        {"path", required_argument, 0, 'p'}, {0,0,0,0}
    };
    int c, idx;
    while ((c = getopt_long(argc, argv, "p:", opts, &idx)) != -1) {
        if (c == 'p') text_path = optarg;
    }

    bp_client_t *cl = NULL;
    if (bp_client_open(&cl) != BP_OK) {
        fprintf(stderr, "client open failed\n"); return 2;
    }
    bp_client_open_session(cl, NULL);

    /* Always print the active node table for context. */
    uint32_t mask_lo = 0, mask_hi = 0;
    if (bp_client_get_active_nodes(cl, &mask_lo, &mask_hi) == BP_OK) {
        printf("[active nodes] mask_lo=0x%08x  mask_hi=0x%08x  =>", mask_lo, mask_hi);
        for (int i = 0; i < 32; i++) if (mask_lo & (1u << i)) printf(" %d", i);
        for (int i = 0; i < 32; i++) if (mask_hi & (1u << i)) printf(" %d", i + 32);
        printf("\n\n");
    }

    if (!text_path) {
        printf("=== LOCAL Identity (bp_client_get_id_local) ===\n");
        bp_id_object_t id;
        int rc = bp_client_get_id_local(cl, &id);
        printf("  rc=%d (%s)\n", rc, bp_strerror(rc));
        if (rc == BP_OK) print_id(&id);
    } else {
        printf("=== REMOTE Identity via OCXcip_GetDeviceIdObject('%s', inst=1) ===\n", text_path);
        bp_id_object_t id;
        int rc = bp_client_get_device_id(cl, text_path, 1, &id);
        printf("  rc=%d (%s)\n", rc, bp_strerror(rc));
        if (rc == BP_OK) print_id(&id);
    }

    bp_client_close(cl);
    return 0;
}

File: python/examples/identity.py
#!/usr/bin/env python3
"""identity — print the Identity object of a CIP device.
Mirrors c/examples/identity.c CLI + output format byte-for-byte.

SPDX-License-Identifier: MIT
"""
import argparse
import sys

import bpclient


def vendor_name(vid: int) -> str:
    return {
        0x0001: "Allen-Bradley (Rockwell)",
        0x0030: "Online Development Inc.",
    }.get(vid, "(unknown)")


def print_id(id_obj: bpclient.Identity) -> None:
    pn = id_obj.product_name
    skip = 0
    while skip < 31 and skip < len(pn) and pn[skip] == 0:
        skip += 1
    take = 0
    while skip + take < 32 and skip + take < len(pn) and pn[skip + take] != 0:
        take += 1
    name = pn[skip:skip + take].decode("ascii", "replace")
    print(f"  Vendor ID       : 0x{id_obj.vendor_id:04x}  ({vendor_name(id_obj.vendor_id)})")
    print(f"  Device Type     : 0x{id_obj.device_type:04x}")
    print(f"  Product Code    : 0x{id_obj.product_code:04x}  ({id_obj.product_code})")
    print(f"  Revision        : {id_obj.major_rev}.{id_obj.minor_rev}")
    print(f"  Status          : 0x{id_obj.status:04x}")
    print(f"  Serial Number   : 0x{id_obj.serial_number:08x}  ({id_obj.serial_number})")
    print(f"  Product Name    : '{name}'")


def main() -> int:
    ap = argparse.ArgumentParser(prog="identity")
    ap.add_argument("--path", default="")
    args = ap.parse_args()

    try:
        client = bpclient.Client()
        client.open()
    except Exception:
        sys.stderr.write("client open failed\n")
        return 2
    try:
        client.open_session()
    except Exception:
        pass

    try:
        lo, hi = client.get_active_nodes()
        parts = []
        for i in range(32):
            if lo & (1 << i):
                parts.append(f" {i}")
        for i in range(32):
            if hi & (1 << i):
                parts.append(f" {i + 32}")
        print(f"[active nodes] mask_lo=0x{lo:08x}  mask_hi=0x{hi:08x}  =>{''.join(parts)}\n")
    except Exception:
        pass

    if not args.path:
        print("=== LOCAL Identity (Client.get_id_local) ===")
        try:
            id_obj = client.get_id_local()
            rc = 0
        except Exception as e:
            id_obj = None
            rc = bpclient.err_code(e)
        print(f"  rc={rc} ({bpclient.strerror(rc)})")
        if id_obj is not None:
            print_id(id_obj)
    else:
        print(f"=== REMOTE Identity via OCXcip_GetDeviceIdObject('{args.path}', inst=1) ===")
        try:
            id_obj = client.get_device_id(args.path, 1)
            rc = 0
        except Exception as e:
            id_obj = None
            rc = bpclient.err_code(e)
        print(f"  rc={rc} ({bpclient.strerror(rc)})")
        if id_obj is not None:
            print_id(id_obj)

    client.close()
    return 0


if __name__ == "__main__":
    sys.exit(main())
